File: programmers/etc/level3/선입_선출_스케줄링.py
def solution(n, cores):
    answer = 0
    if n <= len(cores):
        return n
    
    n -= len(cores)
    left, right = 1, max(cores)*n
    while left+1 < right:
        mid = (left+right)//2
        capacity = 0
        for c in cores:
            capacity += mid//c
        if capacity >= n:
            right = mid
        else:
            left = mid
    
    for c in cores:
        n -= (right-1)//c
    
    for i in range(len(cores)):
        if right%cores[i] == 0:
            n -= 1
            if n == 0:
                return i+1


n, cores = 6, [1,2,3]
print(solution(n, cores)) # 2
   

File: Java/programmers/etc/level3/선입선출스케줄링.java
package CodeTest.Java.programmers.etc.level3;

public class 선입선출스케줄링 {
    public int solution(int n, int[] cores) {
        int c = cores.length;
        
        if (n <= c) return n;
        
        n -= c;
        int _max = 0;
        for (int core : cores) {
            _max = Math.max(_max, core);
        }
        
        int answer = 0;
        int left = 1, right = _max*n;
        while (left <= right) {
            int mid = (left+right)/2;
            int capacity = 0;
            for (int core : cores) {
                capacity += mid/core;
            }
            if (capacity >= n) {
                right = mid-1;
                answer = mid;
            } else {
                left = mid+1;
            }
        }

        for (int core : cores) {
            n -= (answer-1)/core;
        }

        for (int i=0; i<c; i++) {
            if (answer%cores[i] == 0) {
                n -= 1;
                if (n == 0) return i+1;
            }
        }
        
        return 0;
    }
}
